File: src/main/java/gargoyle/xenox/game/scr/menu/config/JControls.java
package gargoyle.xenox.game.scr.menu.config;

import gargoyle.xenox.Xenox;
import gargoyle.xenox.game.sys.Controls;
import gargoyle.xenox.game.sys.Controls.KeyConfig;
import gargoyle.xenox.game.sys.JScreen;
import gargoyle.xenox.util.gui.JKeyButton;
import gargoyle.xenox.util.i18n.Messages;

import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTabbedPane;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;
import java.awt.BorderLayout;
import java.awt.GridLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class JControls extends JDialog {
    public static final String STR_CONTROLS = "controls";
    private static final String STR_CONFIG_PLUS = "+";
    private static final String STR_CONFIG = "config";
    private static final String STR_CONF_ACTION = "action";
    private static final String STR_CONF_DOWN = "down";
    private static final String STR_CONF_LEFT = "left";
    private static final String STR_CONF_RIGHT = "right";
    private static final String STR_CONF_UP = "up";

    public JControls(JScreen parent) {
        super(SwingUtilities.getWindowAncestor(parent), parent.getMessages().get(JOptions.STR_OPTIONS),
                ModalityType.APPLICATION_MODAL);
        setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        init(parent.getMessages(), parent.getControls());
    }

    private static void addConfigPanel(Controls controls, Messages messages, JTabbedPane tabConfigs,
                                       int num, int c, KeyConfig keyConfig) {
        JPanel pnlConfig = new JPanel(new GridLayout(0, 2));
        {
            pnlConfig.add(new JLabel(messages.get(STR_CONF_ACTION)));
            JKeyButton btn = new JKeyButton(keyConfig.getAction());
            btn.addLateActionListener(new KeyActionListener(controls, c, btn));
            pnlConfig.add(btn);
        }
        {
            pnlConfig.add(new JLabel(messages.get(STR_CONF_LEFT)));
            JKeyButton btn = new JKeyButton(keyConfig.getLeft());
            btn.addLateActionListener(new KeyLeftListener(controls, c, btn));
            pnlConfig.add(btn);
        }
        {
            pnlConfig.add(new JLabel(messages.get(STR_CONF_RIGHT)));
            JKeyButton btn = new JKeyButton(keyConfig.getRight());
            btn.addLateActionListener(new KeyRightListener(controls, c, btn));
            pnlConfig.add(btn);
        }
        {
            pnlConfig.add(new JLabel(messages.get(STR_CONF_UP)));
            JKeyButton btn = new JKeyButton(keyConfig.getUp());
            btn.addLateActionListener(new KeyUpListener(controls, c, btn));
            pnlConfig.add(btn);
        }
        pnlConfig.add(new JLabel(messages.get(STR_CONF_DOWN)));
        JKeyButton btn = new JKeyButton(keyConfig.getDown());
        btn.addLateActionListener(new KeyDownListener(controls, c, btn));
        pnlConfig.add(btn);
        tabConfigs.insertTab(String.format("%s %d", messages.get(STR_CONFIG), num), null, pnlConfig, String.valueOf(num),
                num - 1);
    }

    private void init(Messages messages, Controls controls) {
        setLayout(new BorderLayout());
        JTabbedPane tabConfigs = new JTabbedPane();
        int num = 1;
        for (int c = 0; c < controls.getConfigs().size(); c++) {
            KeyConfig keyConfig = controls.getConfigs().get(c);
            addConfigPanel(controls, messages, tabConfigs, num, c, keyConfig);
            num++;
        }
        JPanel pnlNewConfig = new JPanel();
        tabConfigs.addTab(messages.get(STR_CONFIG_PLUS), pnlNewConfig);
        tabConfigs.addChangeListener(evt -> {
            if (tabConfigs.getSelectedComponent() == pnlNewConfig) {
                KeyConfig keyConfig = controls.newConfig(0, 0, 0, 0, 0);
                int c = controls.getConfigs().size();
                tabConfigs.setSelectedIndex(c - 1);
                controls.getConfigs().add(keyConfig);
                addConfigPanel(controls, messages, tabConfigs, c + 1, c, keyConfig);
                pack();
                tabConfigs.setSelectedIndex(c);
            }
        });
        JPanel pnlButtons = new JPanel(new GridLayout(1, 2));
        JButton btnOk = new JButton(messages.get(Xenox.STR_OK));
        btnOk.addActionListener(e -> setVisible(false));
        pnlButtons.add(btnOk);
        JButton btnCancel = new JButton(messages.get(Xenox.STR_CANCEL));
        btnCancel.addActionListener(e -> setVisible(false));
        pnlButtons.add(btnCancel);
        add(tabConfigs, BorderLayout.CENTER);
        add(pnlButtons, BorderLayout.SOUTH);
    }

    private abstract static class BaseKeyListener implements ActionListener {
        final JKeyButton btn;
        final int c;
        final Controls controls;

        BaseKeyListener(Controls controls, int c, JKeyButton btn) {
            this.controls = controls;
            this.c = c;
            this.btn = btn;
        }
    }

    private static final class KeyActionListener extends BaseKeyListener {
        KeyActionListener(Controls controls, int c, JKeyButton btn) {
            super(controls, c, btn);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            controls.getConfigs().get(c).setAction(btn.getKeyCode());
        }
    }

    private static final class KeyDownListener extends BaseKeyListener {
        KeyDownListener(Controls controls, int c, JKeyButton btn) {
            super(controls, c, btn);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            controls.getConfigs().get(c).setDown(btn.getKeyCode());
        }
    }

    private static final class KeyLeftListener extends BaseKeyListener {
        KeyLeftListener(Controls controls, int c, JKeyButton btn) {
            super(controls, c, btn);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            controls.getConfigs().get(c).setLeft(btn.getKeyCode());
        }
    }

    private static final class KeyRightListener extends BaseKeyListener {
        KeyRightListener(Controls controls, int c, JKeyButton btn) {
            super(controls, c, btn);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            controls.getConfigs().get(c).setRight(btn.getKeyCode());
        }
    }

    private static final class KeyUpListener extends BaseKeyListener {
        KeyUpListener(Controls controls, int c, JKeyButton btn) {
            super(controls, c, btn);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            controls.getConfigs().get(c).setUp(btn.getKeyCode());
        }
    }
}

File: src/main/kotlin/gargoyle/xenox/game/scr/menu/config/JControls.kt
package gargoyle.xenox.game.scr.menu.config

import gargoyle.xenox.Xenox
import gargoyle.xenox.game.sys.Controls
import gargoyle.xenox.game.sys.Controls.KeyConfig
import gargoyle.xenox.game.sys.JScreen
import gargoyle.xenox.util.gui.JKeyButton
import gargoyle.xenox.util.i18n.Messages
import java.awt.BorderLayout
import java.awt.GridLayout
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.JButton
import javax.swing.JDialog
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTabbedPane
import javax.swing.SwingUtilities

class JControls(parent: JScreen) : JDialog(
    SwingUtilities.getWindowAncestor(parent), parent.messages[JOptions.STR_OPTIONS],
    ModalityType.APPLICATION_MODAL
) {
    init {
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        init(parent.messages, parent.controls)
    }

    private fun init(messages: Messages, controls: Controls) {
        layout = BorderLayout()
        val tabConfigs = JTabbedPane()
        var num = 1
        for (c in controls.configs.indices) {
            val keyConfig = controls.configs[c]
            addConfigPanel(controls, messages, tabConfigs, num, c, keyConfig)
            num++
        }
        val pnlNewConfig = JPanel()
        tabConfigs.addTab(messages[STR_CONFIG_PLUS], pnlNewConfig)
        tabConfigs.addChangeListener {
            if (tabConfigs.selectedComponent === pnlNewConfig) {
                val keyConfig = controls.newConfig(0, 0, 0, 0, 0)
                val c = controls.configs.size
                tabConfigs.selectedIndex = c - 1
                controls.configs.add(keyConfig)
                addConfigPanel(controls, messages, tabConfigs, c + 1, c, keyConfig)
                pack()
                tabConfigs.selectedIndex = c
            }
        }
        val pnlButtons = JPanel(GridLayout(1, 2))
        val btnOk = JButton(messages[Xenox.STR_OK])
        btnOk.addActionListener { isVisible = false }
        pnlButtons.add(btnOk)
        val btnCancel = JButton(messages[Xenox.STR_CANCEL])
        btnCancel.addActionListener { isVisible = false }
        pnlButtons.add(btnCancel)
        add(tabConfigs, BorderLayout.CENTER)
        add(pnlButtons, BorderLayout.SOUTH)
    }

    private abstract class BaseKeyListener(
        val controls: Controls,
        val c: Int,
        val btn: JKeyButton
    ) : ActionListener

    private class KeyActionListener(controls: Controls, c: Int, btn: JKeyButton) :
        BaseKeyListener(controls, c, btn) {
        override fun actionPerformed(e: ActionEvent) {
            controls.configs[c].setAction(btn.keyCode)
        }
    }

    private class KeyDownListener(controls: Controls, c: Int, btn: JKeyButton) :
        BaseKeyListener(controls, c, btn) {
        override fun actionPerformed(e: ActionEvent) {
            controls.configs[c].setDown(btn.keyCode)
        }
    }

    private class KeyLeftListener(controls: Controls, c: Int, btn: JKeyButton) :
        BaseKeyListener(controls, c, btn) {
        override fun actionPerformed(e: ActionEvent) {
            controls.configs[c].setLeft(btn.keyCode)
        }
    }

    private class KeyRightListener(controls: Controls, c: Int, btn: JKeyButton) :
        BaseKeyListener(controls, c, btn) {
        override fun actionPerformed(e: ActionEvent) {
            controls.configs[c].setRight(btn.keyCode)
        }
    }

    private class KeyUpListener(controls: Controls, c: Int, btn: JKeyButton) :
        BaseKeyListener(controls, c, btn) {
        override fun actionPerformed(e: ActionEvent) {
            controls.configs[c].setUp(btn.keyCode)
        }
    }

    companion object {
        const val STR_CONTROLS = "controls"
        private const val STR_CONFIG_PLUS = "+"
        private const val STR_CONFIG = "config"
        private const val STR_CONF_ACTION = "action"
        private const val STR_CONF_DOWN = "down"
        private const val STR_CONF_LEFT = "left"
        private const val STR_CONF_RIGHT = "right"
        private const val STR_CONF_UP = "up"
        private fun addConfigPanel(
            controls: Controls, messages: Messages, tabConfigs: JTabbedPane,
            num: Int, c: Int, keyConfig: KeyConfig
        ) {
            val pnlConfig = JPanel(GridLayout(0, 2))
            run {
                pnlConfig.add(JLabel(messages[STR_CONF_ACTION]))
                val btn = JKeyButton(keyConfig.getAction())
                btn.addLateActionListener(KeyActionListener(controls, c, btn))
                pnlConfig.add(btn)
            }
            run {
                pnlConfig.add(JLabel(messages[STR_CONF_LEFT]))
                val btn = JKeyButton(keyConfig.getLeft())
                btn.addLateActionListener(KeyLeftListener(controls, c, btn))
                pnlConfig.add(btn)
            }
            run {
                pnlConfig.add(JLabel(messages[STR_CONF_RIGHT]))
                val btn = JKeyButton(keyConfig.getRight())
                btn.addLateActionListener(KeyRightListener(controls, c, btn))
                pnlConfig.add(btn)
            }
            run {
                pnlConfig.add(JLabel(messages[STR_CONF_UP]))
                val btn = JKeyButton(keyConfig.getUp())
                btn.addLateActionListener(KeyUpListener(controls, c, btn))
                pnlConfig.add(btn)
            }
            pnlConfig.add(JLabel(messages[STR_CONF_DOWN]))
            val btn = JKeyButton(keyConfig.getDown())
            btn.addLateActionListener(KeyDownListener(controls, c, btn))
            pnlConfig.add(btn)
            tabConfigs.insertTab(
                String.format("%s %d", messages[STR_CONFIG], num), null, pnlConfig, num.toString(),
                num - 1
            )
        }
    }
}
